### PR TITLE
Try some optimizations

### PR DESCRIPTION
**Pull request by odersky**
This branch tries out some low-level optimizations. It also contains `Fix:` commits which are fixes of problems discovered.

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test performance please: a1ce848

**Comment by liufengyun** [comment]
@odersky it's possible to give multiple commits separated by space.

File: tests/run/lst/Lst.scala [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+    case elems: Array[AnyRef] => def elem(i: Int) = elems(i).asInstanceOf[T]
+      elem(n)
+    case elem: T @unchecked =>
+      if (n == 0) elem else throw new IndexOutOfBoundsException(n.toString)
+  }
+
+  def head: T = apply(0)
+  def last: T = apply(length - 1)
+
+  def headOr[U >: T](alt: => U): U = if (length == 0) alt else head
```
Comment (allanrenucci) [inline tests/run/lst/Lst.scala]:
```scala
if (isEmpty) alt else head
```

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
Rebased sur #4464.

**Comment by odersky** [comment]
Test performance please

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
The fixes in this PR have been integrated in master. As far as optimizations go, the small ones did not help and for `Lst` we need a fresh start because it gets too messy to mix `Lst` and `List`. The idea is to globally replace `List` by `Lst` in all the compiler, including the backend, either by changing the implementation of `List` in the standard library or by installing `Lst`, `Lst.Empty` and `Lst.::` as aliases for the `List` operations in a global package object.

The updated `Lst` is part of #4490

**Pull request closed without merging**